src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -290,24 +290,40 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
                 var projectId = documentId.ProjectId;
                 var dte = (EnvDTE.DTE)_serviceProvider.GetService(typeof(SDTE));
                 var dteProject = _workspace.TryGetDTEProject(projectId);
-                if (dteProject != null)
+                var projectGuid = _workspace.GetProjectGuid(projectId);
+                if (dteProject != null && projectGuid != Guid.Empty)
                 {
                     var undoManager = _editorAdaptersFactoryService.TryGetUndoManager(
                         workspace, documentId, cancellationToken);
 
                     return TryInstallAndAddUndoAction(
-                        source, packageName, versionOpt, includePrerelease, dte, dteProject, undoManager);
+                        source, packageName, versionOpt, includePrerelease, projectGuid, dte, dteProject, undoManager);
                 }
             }
 
             return false;
         }
 
+        private bool IsPackageInstalled(Guid projectGuid, string packageName)
+        {
+            var cancellationToken = CancellationToken.None;
+
+            return this.ThreadingContext.JoinableTaskFactory.Run(() =>
+            {
+                return this.PerformNuGetProjectServiceWorkAsync(async nugetService =>
+                {
+                    var installedPackagesMap = await GetInstalledPackagesMapAsync(nugetService, projectGuid, cancellationToken).ConfigureAwait(false);
+                    return installedPackagesMap.ContainsKey(packageName);
```

Comment:
If you want 100% parity with the old service, you'd only consider top level packages. 

It really depends how you'd use the package, a decision that roslyn can't know. 

I don't expect  too many side-effects either way, but preserving the current behavior might be something you'd prefer.